File: voice-memo-watch/watchos/VoiceMemoBot/AppConfig.swift
import Foundation

enum AppConfig {
    /// Base URL of the VoiceMemoBot backend. For simulator testing point this
    /// at your dev machine (e.g. http://192.168.x.x:8000); in production at
    /// the deployed API. Must be HTTPS for App Store builds (ATS).
    static let apiBaseURL = URL(string: "http://localhost:8000")!
}

File: voice-memo-watch/wearos/app/src/main/java/com/voicememobot/wear/AppConfig.kt
package com.voicememobot.wear

/**
 * Where the VoiceMemoBot backend lives. For the emulator use 10.0.2.2 to
 * reach the host machine; on a real watch set your machine's LAN IP or the
 * deployed Render URL (HTTPS).
 */
object AppConfig {
    const val API_BASE_URL = "http://10.0.2.2:8000"
}
